porch/func/internal/podevaluator.go

```diff
@@ -64,40 +68,312 @@ func NewPodEvaluator(namespace, wrapperServerImage string) (Evaluator, error) {
+					if host == pod.Status.PodIP {
+						delete(pcm.cache, image)
+					}
+				}
+
+				go func(po corev1.Pod) {
+					klog.Infof("deleting pod %v/%v", po.Namespace, po.Name)
+					err := pcm.podManager.kubeClient.Delete(context.Background(), &po)
+					if err != nil {
+						klog.Warningf("unable to delete pod %v/%v: %w", po.Namespace, po.Name, err)
+					}
+				}(podList.Items[i])
+			}
+		}
+	}
+}
+
+type podManager struct {
+	// kubeClient is the kubernetes client
+	kubeClient client.Client
+	// namespace holds the namespace where the executors run
+	namespace string
+	// wrapperServerImage is the image name of the wrapper server
+	wrapperServerImage string
+
+	// podReadyCh is a channel to send the grpc client information.
+	// podCacheManager receives from this channel.
+	podReadyCh chan *imagePodAndGRPCClient
+
+	// entrypointCache is a cache of image name to entrypoint.
+	// Only podManager is allowed to touch this cache.
+	// Its underlying type is map[string][]string.
+	entrypointCache sync.Map
+}
+
+func (pm *podManager) getFuncEvalPodClient(ctx context.Context, image string) {
+	c, err := func() (*podAndGRPCClient, error) {
+		podKey, err := pm.retrieveOrCreatePod(ctx, image)
+		if err != nil {
+			return nil, err
+		}
+		podIP, err := pm.podIpIfRunningAndReady(ctx, podKey)
+		if err != nil {
+			return nil, err
+		}
+		if podIP == "" {
+			return nil, fmt.Errorf("pod %s/%s did not have podIP", podKey.Namespace, podKey.Name)
+		}
+		address := net.JoinHostPort(podIP, defaultWrapperServerPort)
+		cc, err := grpc.Dial(address, grpc.WithTransportCredentials(insecure.NewCredentials()))
+		if err != nil {
+			return nil, fmt.Errorf("failed to dial grpc function evaluator on %q for pod %s/%s: %w", address, podKey.Namespace, podKey.Name, err)
+		}
+		return &podAndGRPCClient{
+			pod:        podKey,
+			grpcClient: cc,
+		}, err
+	}()
+	pm.podReadyCh <- &imagePodAndGRPCClient{
+		image:            image,
+		podAndGRPCClient: c,
+		err:              err,
+	}
+}
+
+func (pm *podManager) imageEntrypoint(image string) ([]string, error) {
+	// Create pod otherwise.
+	var entrypoint []string
+	ref, err := name.ParseReference(image)
 	if err != nil {
 		return nil, err
 	}
 	img, err := remote.Image(ref, remote.WithAuthFromKeychain(authn.DefaultKeychain))
 	if err != nil {
 		return nil, err
 	}
-	// TODO: cache the config file
 	cf, err := img.ConfigFile()
 	if err != nil {
 		return nil, err
 	}
 
 	cfg := cf.Config
-	var fnCmd []string
+	// TODO: to handle all scenario, we should follow https://docs.docker.com/engine/reference/builder/#understand-how-cmd-and-entrypoint-interact.
 	if len(cfg.Entrypoint) != 0 {
-		fnCmd = cfg.Entrypoint
+		entrypoint = cfg.Entrypoint
 	} else {
-		fnCmd = cfg.Cmd
+		entrypoint = cfg.Cmd
+	}
+	pm.entrypointCache.Store(image, entrypoint)
+	return entrypoint, nil
+}
+
+func (pm *podManager) retrieveOrCreatePod(ctx context.Context, image string) (client.ObjectKey, error) {
+	// Try to retrieve the pod first.
```

Comment:
the language referring to ordering (here in few other places) seems out-of-date since (I assume) refactoring created independent functions.

Maybe `// Try to retrieve the pod.`